src/physlr-map.cc

```diff
@@ -0,0 +1,561 @@
+#include "tsl/robin_map.h"
+#include "tsl/robin_set.h"
+
+#include <cfenv>
+#include <cstdint>
+#include <cstdlib>
+#include <fstream>
+#include <getopt.h>
+#include <iomanip>
+#include <iostream>
+#include <sstream>
+#include <string>
+#include <thread>
+#include <vector>
+
+#include <boost/functional/hash.hpp>
+
+#if _OPENMP
+#include <omp.h>
+#endif
+
+/**
+ * Global variables that are mostly constant for the duration of the
+ * execution of the program.
+ */
+namespace opt {
+static unsigned threads = 1;
+static unsigned scoreThreshold = 10;
+static unsigned mapPositions = 10;
+} // namespace opt
+
+uint64_t max = std::numeric_limits<std::uint64_t>::max();
+
+#define PROGRAM "physlr-map"
+#define PACKAGE_NAME "physlr"
+#define GIT_REVISION "pre-autotools"
+
+static uint64_t
+memory_usage()
+{
+	int mem = 0;
+	std::ifstream proc("/proc/self/status");
+	for (std::string s; std::getline(proc, s);) {
+		if (s.substr(0, 6) == "VmSize") {
+			std::stringstream convert(s.substr(s.find_last_of('\t'), s.find_last_of('k') - 1));
+			if (!(convert >> mem)) {
+				return 0;
+			}
+			return mem;
+		}
+	}
+	return mem;
+}
+
+using BarcodeID = uint32_t;
+using Minimizer = uint64_t;
+using pair = std::pair<uint64_t, uint64_t>;
+using barcodeToMinimizer_t = tsl::robin_map<std::string, std::vector<Minimizer>>;
+using path_t = std::vector<std::string>;
+using paths_t = std::vector<path_t>;
+using minimizerToPos_t = tsl::robin_map<Minimizer, tsl::robin_set<pair, boost::hash<pair>>>;
+
+static void
+printVersion()
+{
+	const char VERSION_MESSAGE[] =
+	    PROGRAM " (" PACKAGE_NAME ") " GIT_REVISION "\n"
+	            "Written by Johnathan Wong.\n"
+	            "\n"
+	            "Copyright 2020 Canada's Michael Smith Genome Science Centre\n";
+	std::cerr << VERSION_MESSAGE << std::endl;
+	exit(EXIT_SUCCESS);
+}
+
+static void
+printHelpDialog()
+{
+	static const char dialog[] =
+	    "Usage: physlr-map [OPTION]... [TargetPATHS.path] [TargetMINIMIZERS.tsv] "
+	    "[QueryMINIMIZERS.tsv]\n"
+	    "Map sequences to a physical map.\n"
+	    "  -t, --threads     threads [1]\n"
+	    "  -n, --scoreThreshold     minimum score to map [10]\n"
+	    "  -m, --mapPositions     number of positions used to orient [10]\n"
+	    "  -v         enable verbose output\n"
+	    "  --version     Print version\n"
+	    "  --help     display this help and exit\n"
+	    "Report bugs to <jowong@bcgsc.ca>.";
+	std::cerr << dialog << std::endl;
+	exit(EXIT_SUCCESS);
+}
+
+static void
+printErrorMsg(const std::string& progname, const std::string& msg)
+{
+	std::cerr << progname << ": " << msg << "\nTry '" << progname
+	          << " --help' for more information.\n";
+}
+
+uint64_t
+lowerMedian(std::vector<uint64_t> scores)
+{
+	uint64_t size = scores.size();
+
+	if (size == 0) {
+		return 0;
+	}
+
+	sort(scores.begin(), scores.end());
```

Comment:
Just double checking because I recall us discussing for python -- it is OK that this sorts in place?

Reply:
`lowerMedian(std::vector<uint64_t> scores)` this is not passing by reference so it should be ok.